tests/qos/test_pfc_pause.py

```diff
@@ -0,0 +1,396 @@
+
+from ansible_host import AnsibleHost
+from qos_fixtures import *
+from qos_helpers import ansible_stdout_to_str, eos_to_linux_intf, get_active_intfs, get_addrs_in_subnet, get_neigh_ip,check_mac_table, get_mac, config_intf_ip_mac
+import pytest
+import os
+import time
+import re
+import struct
+import random
+import ipaddress
+
+PFC_GEN_FILE = 'pfc_gen.py'
+PFC_GEN_FILE_RELATIVE_PATH = '../../ansible/roles/test/files/helpers/pfc_gen.py'
+PFC_GEN_FILE_DEST = '~/pfc_gen.py'
+PFC_PKT_COUNT = 1000000000
+
+PTF_FILE_RELATIVE_PATH = '../../ansible/roles/test/files/ptftests/pfc_pause_test.py'
+PTF_FILE_DEST = '~/pfc_pause_test.py'
+PTF_PKT_COUNT = 50
+PTF_PKT_INTVL_SEC = 0.1
+PTF_PASS_RATIO_THRESH = 0.75
+
+""" Maximum number of interfaces to test on a DUT """
+MAX_TEST_INTFS_COUNT = 4
+         
+def atoi(text):
+    return int(text) if text.isdigit() else text
+
+def natural_keys(text):
+    return [atoi(c) for c in re.split(r'(\d+)', text)]
+
+def setup_testbed(ansible_adhoc, testbed, leaf_fanouts):
+    """
+    @Summary: Set up the testbed, including:
+    (1) copying the PFC generator to the leaf fanout switches,
+    (2) stopping PFC storm at the leaf fanout switches,
+    (3) copying the PTF script to the PTF container.
+    @param ansible_adhoc: Fixture provided by the pytest-ansible package. Source of the various device objects. It is
+    mandatory argument for the class constructors.
+    @param testbed: Testbed information
+    @param leaf_fanouts: Leaf fanout switches
+    """
+    
+    """ Copy the PFC generator to leaf fanout switches """
+    for peer_device in leaf_fanouts:
+        peerdev_ans = AnsibleHost(ansible_adhoc, peer_device)
+        cmd = "sudo kill -9 $(pgrep -f %s) </dev/null >/dev/null 2>&1 &" % (PFC_GEN_FILE)
+        peerdev_ans.shell(cmd)
+        file_src = os.path.join(os.path.dirname(__file__), PFC_GEN_FILE_RELATIVE_PATH)
+        peerdev_ans.copy(src=file_src, dest=PFC_GEN_FILE_DEST, force=True)
+   
+    """ Stop PFC storm at the leaf fanout switches """
+    for peer_device in leaf_fanouts:
+        peerdev_ans = AnsibleHost(ansible_adhoc, peer_device)
+        cmd = "sudo kill -9 $(pgrep -f %s) </dev/null >/dev/null 2>&1 &" % (PFC_GEN_FILE) 
+        peerdev_ans.shell(cmd) 
+                       
+    """ Remove existing python scripts on PTF """
+    ptf_hostname = testbed['ptf']
+    ptf_ans = AnsibleHost(ansible_adhoc, ptf_hostname)
+    result = ptf_ans.find(paths=['~/'], patterns="*.py")['files']
+    files = [ansible_stdout_to_str(x['path']) for x in result]
+    
+    for file in files:
+        ptf_ans.file(path=file, mode="absent")
+
+    """ Copy the PFC test script to the PTF container """  
+    file_src = os.path.join(os.path.dirname(__file__), PTF_FILE_RELATIVE_PATH)
+    ptf_ans.copy(src=file_src, dest=PTF_FILE_DEST, force=True)
+    
+def run_test_t0(ansible_adhoc, 
```

Comment:
I would suggest to break down this function into smaller functions. vlan setup should be in its own function, mac population steps should be in another one. start and stop pfc storm should be a separate function These are again part of the setup that needs to be done before the test run. So they should be called by a setup function. The run test should start the pfc storm, run the traffic and verify functionality